cmd/influx/backup.go

```diff
@@ -1,97 +1,328 @@
+	genericCLIOpts
+	*globalFlags
+
+	bucketID   string
+	bucketName string
+	org        organization
+	path       string
+
+	manifest influxdb.Manifest
+	baseName string
+
+	backupService *http.BackupService
+	kvStore       *bolt.KVStore
+	kvService     *kv.Service
+	metaClient    *meta.Client
+
+	logger *zap.Logger
+}
 
+func newCmdBackupBuilder(f *globalFlags, opts genericCLIOpts) *cmdBackupBuilder {
+	return &cmdBackupBuilder{
+		genericCLIOpts: opts,
+		globalFlags:    f,
+	}
+}
+
+func (b *cmdBackupBuilder) cmdBackup() *cobra.Command {
+	cmd := b.newCmd("backup", b.backupRunE)
+	b.org.register(cmd, true)
+	cmd.Flags().StringVar(&b.bucketID, "bucket-id", "", "The ID of the bucket to backup")
+	cmd.Flags().StringVarP(&b.bucketName, "bucket", "b", "", "The name of the bucket to backup")
+	cmd.Use = "backup [flags] path"
+	cmd.Args = func(cmd *cobra.Command, args []string) error {
+		if len(args) == 0 {
+			return fmt.Errorf("must specify output path")
+		} else if len(args) > 1 {
+			return fmt.Errorf("too many args specified")
+		}
+		b.path = args[0]
+		return nil
+	}
+	cmd.Short = "Backup database"
+	cmd.Long = `
+Backs up InfluxDB to a directory.
+
+Examples:
+	# backup all data
+	influx backup /path/to/backup
+`
 	return cmd
 }
 
-var backupFlags struct {
-	Path string
+func (b *cmdBackupBuilder) manifestPath() string {
+	return fmt.Sprintf("%s.manifest", b.baseName)
 }
 
-func newBackupService() (influxdb.BackupService, error) {
+func (b *cmdBackupBuilder) kvPath() string {
+	return fmt.Sprintf("%s.bolt", b.baseName)
+}
+
+func (b *cmdBackupBuilder) shardPath(id uint64) string {
+	return fmt.Sprintf("%s.s%d", b.baseName, id) + ".tar.gz"
+}
+
+func (b *cmdBackupBuilder) backupRunE(cmd *cobra.Command, args []string) (err error) {
+	ctx := context.Background()
+
+	// Create top level logger
+	logconf := influxlogger.NewConfig()
+	if b.logger, err = logconf.New(os.Stdout); err != nil {
+		return err
+	}
+
+	// Determine a base
+	b.baseName = time.Now().UTC().Format(influxdb.BackupFilenamePattern)
+
+	// Ensure directory exsits.
+	if err := os.MkdirAll(b.path, 0777); err != nil {
+		return err
+	}
+
 	ac := flags.config()
-	return &http.BackupService{
+	b.backupService = &http.BackupService{
 		Addr:  ac.Host,
 		Token: ac.Token,
-	}, nil
-}
+	}
 
-func backupF(cmd *cobra.Command, args []string) error {
-	ctx := context.Background()
+	// Back up Bolt database to file.
+	if err := b.backupKVStore(ctx); err != nil {
+		return err
+	}
 
-	if backupFlags.Path == "" {
-		return fmt.Errorf("must specify path")
+	// Open bolt DB.
+	boltClient := bolt.NewClient(b.logger)
+	boltClient.Path = filepath.Join(b.path, b.kvPath())
+	if err := boltClient.Open(ctx); err != nil {
+		return err
 	}
+	defer boltClient.Close()
+
+	// Open meta store so we can iterate over meta data.
+	b.kvStore = bolt.NewKVStore(b.logger, filepath.Join(b.path, b.kvPath()))
+	b.kvStore.WithDB(boltClient.DB())
+	b.kvService = kv.NewService(b.logger, b.kvStore, kv.ServiceConfig{})
```

Comment:
The Bolt file may be mutated if the database has pending migrations:

https://github.com/influxdata/influxdb/blob/d8a10e531130a70461bf21cc48b5cd3ec3eb0e9e/bolt/bbolt.go#L72

Given it is opened concurrently with the running `influxd` process, would this not result in potential data corruption?

Reply:
This is opening the copy of the bolt database that's been backed up here: https://github.com/influxdata/influxdb/pull/19864/files/0ebae228bc8a4938509b766eb650335deaa8f8eb#diff-7991a9e9385fcbc88c634be0be2409636caa243ff1ac0afa38dfbdb7eda6273eR115-R118

Also, bolt uses an advisory lock so it would hang if it tried to open the same bolt database as the server.

Reply:
Sorry, I totally misread the source path!